### PR TITLE
Enabling MIRI Photometry

### LINKED ISSUE #313: [Enhancement]: Photometry Routines
As we've been discussing for a while, we need photometry methods for the simultaneous photometry from NIRCam and from specific photometry observations (e.g. Laura and I (through Tom Greene) each have TRAPPIST-1 photometry observations coming from MIRI). I believe @sebastian-zieba has already done most of the work of repurposing previous photometry code (POET based code I believe).

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### LINKED ISSUE #490: [Enhancement]: Safer skyin + skyout aperture specification
### Instrument

NIRCam (Stages 1-3), MIRI (Stages 1-3)

### What is your suggestion?

As of right now, for photometric background annulus selection is specified using skyin and skyout as the inner and outer radii of the annulus. This leaves open the possibility of a user specifying an inside-out annulus where skyout is smaller than skyin. It also makes it hard to specify a range of skyin and skyout sizes that the code should automatically try as some subset of the skyout array might be interior to some subset of the skyin array.

In my opinion, a better solution would be to specify skyin and skywidth, so to have an annulus spanning 15-25 pixels, you'd specify skyin=15, skywidth=10. That'll easily allow you to try a range of aperture sizes and inner radii without having to have any code to check that skyout is larger than skyin.

### Error traceback output

_No response_

### What operating system are you using?

_No response_

### What version of Python are you running?

_No response_

### What Python packages do you have installed?

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by taylorbell57**
This PR enables the use of Eureka! for MIRI photometry.

In particular, this PR:
- Edits miri.py's read function to permit photometry files
- Adds the required flag_bg_phot function to miri.py (just calls NIRCam's function)
- Edits miri.py's read function to better handle simulated data without INTSTART and INTEND header values in the FITS files
- Fixes some minor bugs with plots 3503 (not writing over previous plot and consuming excessive RAM; also wasn't being limited by meta.nplots), 3306 & 3504 (hardcoded vmin and vmax weren't good for all datasets), 3505 (run for all integrations and not limited by meta.nplots)
- Changes the behaviour of meta.skyin and meta.skyout if they are lists. Previously, there were no protections against skyin being larger than skyout, so the new method assumes that there is one skyin value for each skyout value. In otherwords, if skyin=[10,11] and skyout=[15,16], previously the code would've tried 10-15, 10-16, 11-15, and 11-16 annuli while now it will just do 10-15 and 10-16. That's important in the case of skyin=[10,15,20], skyout=[15,20,25], where the old algorithm could do some pretty dangerous things.
- Makes running the b2f.convert_to_e function optional (but run by default if the new meta.convert_to_e parameter is missing). This is important to calculate flux-calibrated stellar photometry.
- Allows a manual first-guess of the centroid position rather than the currently implemented full-frame centroid fit. For the NIRCam photometry, the full-frame fit isn't too far off since the PSF is huge and the star is bright, but for MIRI the star is very small and not much brighter than the background (in some filters at least) and there's also significant non-uniformities in the huge full-frame images which make the full-frame centroiding even worse. Users can now pass in meta.ctr_guess with the manually generated x,y guess for the centroid (using the pixel indices before xwindow and ywindow are applied).
- Adds a new ECF template for MIRI photometry
- Documents all added/changed ECF parameters

This PR also:
- Resolves #490 (Using skyin and skywidth is probably the safer choice though, and if you agree I can switch to the approach I suggested in that issue)
- Resolves #313 (Extends photometry support to MIRI, the second and only other JWST instrument that allows photometry to my knowledge)

I think @sebastian-zieba also had a MIRI photometry branch under development. Depending on how close that is to being ready for a PR, we can either:
1. Merge this PR first and then merge in Sebastian's code when it's ready
2. Merge Sebastian's PR into this branch and then merge this branch into main

### CODE REVIEW

File: src/eureka/S3_data_reduction/s3_reduce.py
Comment (kevin218) [inline src/eureka/S3_data_reduction/s3_reduce.py:440]: I'm not keen on requiring a fixed `position` when a guess is provided.  If the user supplies a guess centroid, they should also have the option to feed the guess into `centerdriver`.

File: docs/source/ecf.rst
Comment (taylorbell57) [reply docs/source/ecf.rst:272]: This isn't a required setting (by default is set to true if not present in the ECF) and is only needed if you want to do flux-calibrated photometry. I can add it to the NIRCam and MIRI photometry ECFs if you want, but I really don't think it should be added to the spectroscopy ECFs since it's likely not overly useful there and more likely to cause confusion than anything else

File: src/eureka/S3_data_reduction/miri.py
Comment (taylorbell57) [reply src/eureka/S3_data_reduction/miri.py:82]: I agree this is a bit clunky, but I think it's good enough for now. One complication to keep in mind though is that the exact wavelength is dependent on the stellar spectrum convolved with the filter's throughput.

File: src/eureka/S3_data_reduction/s3_reduce.py
Comment (taylorbell57) [reply src/eureka/S3_data_reduction/s3_reduce.py:440]: Well the guess just replaces the first round of centroiding which was done on the entire frame. For MIRI, centroiding on the entire frame is a really bad idea because the frame is so large and there can easily be multiple sources, and even if there is only one source in the field the centroid gets strongly affected by the very non-uniform background.

To be clear, a second round of centroiding is still done in lines 451-464 in the exact same was as was done before (using a smaller subarray around the guess), just using the manually guessed value instead of the full-frame automatic centroiding. And if no guess is provided, then the code runs the exact same as before.

Comment (kevin218) [reply src/eureka/S3_data_reduction/s3_reduce.py:440]: Ah, ok.  Thanks for clarifying.  Now it makes sense.

File: docs/source/ecf.rst
Comment (kevin218) [reply docs/source/ecf.rst:272]: Please add it to the spectroscopic ECFs.  Today, I've been trying to generate a calibrated stellar spectrum with G395 and has to comment out the conversion line in Stage 3.

File: demos/JWST/S3_miri_photometry_template.ecf
Comment (taylorbell57) [reply demos/JWST/S3_miri_photometry_template.ecf:18]: The nircam photometry file will also need to be updated

**Comment by taylorbell57** [comment]
Alight, I think I've addressed all your comments (aside from the edits that need to wait for #502 to be merged).

I do, however, want your opinion on whether I should continue using `skyin` and `skyout` as is currently done or instead use `skyin` and `skywidth` to make it impossible to have `skyout < skyin` and to make it a bit easier to loop over different aperture options (using an n*m grid of `skyin` and `skywidth` values instead of having to specify one `skyout` for each `skyin`)

**Review by kevin218 — COMMENTED** [review]
Changes look good.  I think using `skywidth` makes the most sense.

File: demos/JWST/S3_miri_photometry_template.ecf
Comment (taylorbell57) [reply demos/JWST/S3_miri_photometry_template.ecf:18]: So the photometry code's `flag_bg_phot` function currently does sigma clipping along the time axis for each individual background pixel independently. As a result, your full-frame code isn't exactly interchangeable with the photometry code. The `flag_bg` boolean is also a on/off switch, and if set False no background sigma clipping is done. It's not completely clear to me how to integrate your new code if that's what you're wanting to do, but as it stands your PR #502 did not make any changes to the photometry code that mean I need to change this new ECF

Comment (kevin218) [reply demos/JWST/S3_miri_photometry_template.ecf:18]: I can look into implementing full-frame sigma clipping for photometry once both PRs are merged.

**Review by kevin218 — APPROVED** [review]

**Pull request merged**
